// @aws-amplify/data-schema/dist/cjs/a.js
'use strict';

Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.ai = exports.generation = exports.conversation = exports.ipAddress = exports.url = exports.phone = exports.json = exports.email = exports.timestamp = exports.datetime = exports.time = exports.date = exports.boolean = exports.float = exports.integer = exports.string = exports.id = exports.belongsTo = exports.hasMany = exports.hasOne = exports.subscription = exports.mutation = exports.query = exports.enum = exports.customType = exports.ref = exports.model = exports.combine = exports.schema = void 0;
const ModelSchema_1 = require("./ModelSchema");
Object.defineProperty(exports, "schema", { enumerable: true, get: function () { return ModelSchema_1.schema; } });
const CombineSchema_1 = require("./CombineSchema");
Object.defineProperty(exports, "combine", { enumerable: true, get: function () { return CombineSchema_1.combine; } });
const ModelType_1 = require("./ModelType");
Object.defineProperty(exports, "model", { enumerable: true, get: function () { return ModelType_1.model; } });
const ModelField_1 = require("./ModelField");
Object.defineProperty(exports, "id", { enumerable: true, get: function () { return ModelField_1.id; } });
Object.defineProperty(exports, "string", { enumerable: true, get: function () { return ModelField_1.string; } });
Object.defineProperty(exports, "integer", { enumerable: true, get: function () { return ModelField_1.integer; } });
Object.defineProperty(exports, "float", { enumerable: true, get: function () { return ModelField_1.float; } });
Object.defineProperty(exports, "boolean", { enumerable: true, get: function () { return ModelField_1.boolean; } });
Object.defineProperty(exports, "date", { enumerable: true, get: function () { return ModelField_1.date; } });
Object.defineProperty(exports, "time", { enumerable: true, get: function () { return ModelField_1.time; } });
Object.defineProperty(exports, "datetime", { enumerable: true, get: function () { return ModelField_1.datetime; } });
Object.defineProperty(exports, "timestamp", { enumerable: true, get: function () { return ModelField_1.timestamp; } });
Object.defineProperty(exports, "email", { enumerable: true, get: function () { return ModelField_1.email; } });
Object.defineProperty(exports, "json", { enumerable: true, get: function () { return ModelField_1.json; } });
Object.defineProperty(exports, "phone", { enumerable: true, get: function () { return ModelField_1.phone; } });
Object.defineProperty(exports, "url", { enumerable: true, get: function () { return ModelField_1.url; } });
Object.defineProperty(exports, "ipAddress", { enumerable: true, get: function () { return ModelField_1.ipAddress; } });
const RefType_1 = require("./RefType");
Object.defineProperty(exports, "ref", { enumerable: true, get: function () { return RefType_1.ref; } });
const ModelRelationshipField_1 = require("./ModelRelationshipField");
Object.defineProperty(exports, "hasOne", { enumerable: true, get: function () { return ModelRelationshipField_1.hasOne; } });
Object.defineProperty(exports, "hasMany", { enumerable: true, get: function () { return ModelRelationshipField_1.hasMany; } });
Object.defineProperty(exports, "belongsTo", { enumerable: true, get: function () { return ModelRelationshipField_1.belongsTo; } });
const CustomType_1 = require("./CustomType");
Object.defineProperty(exports, "customType", { enumerable: true, get: function () { return CustomType_1.customType; } });
const EnumType_1 = require("./EnumType");
Object.defineProperty(exports, "enum", { enumerable: true, get: function () { return EnumType_1.enumType; } });
const CustomOperation_1 = require("./CustomOperation");
Object.defineProperty(exports, "query", { enumerable: true, get: function () { return CustomOperation_1.query; } });
Object.defineProperty(exports, "mutation", { enumerable: true, get: function () { return CustomOperation_1.mutation; } });
Object.defineProperty(exports, "subscription", { enumerable: true, get: function () { return CustomOperation_1.subscription; } });
Object.defineProperty(exports, "generation", { enumerable: true, get: function () { return CustomOperation_1.generation; } });
const Handler_1 = require("./Handler");
Object.defineProperty(exports, "handler", { enumerable: true, get: function () { return Handler_1.handler; } });
const ConversationType_1 = require("./ai/ConversationType");
Object.defineProperty(exports, "conversation", { enumerable: true, get: function () { return ConversationType_1.conversation; } });
const ModelType_2 = require("./ai/ModelType");
const ai = {
    model: ModelType_2.model,
};
exports.ai = ai;
